File: src/FSharpVSPowerTools/RenameCommandFilterProvider.cs
using System.Diagnostics;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio;
using Microsoft.VisualStudio.Editor;
using Microsoft.VisualStudio.OLE.Interop;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.TextManager.Interop;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using FSharpVSPowerTools.ProjectSystem;
using FSharpVSPowerTools.Refactoring;
using Microsoft.VisualStudio.Text;

namespace FSharpVSPowerTools
{
    [Export(typeof(IVsTextViewCreationListener))]
    [ContentType("F#")]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    internal class RenameCommandFilterProvider : IVsTextViewCreationListener
    {
        [Import]
        internal IVsEditorAdaptersFactoryService editorFactory = null;

        [Import]
        internal ITextDocumentFactoryService textDocumentFactoryService = null;

        [Import]
        internal VSLanguageService fsharpVsLanguageService = null;

        [Import(typeof(SVsServiceProvider))]
        internal System.IServiceProvider serviceProvider = null;

        [Import(typeof(ProjectFactory))]
        internal ProjectFactory projectFactory = null;

        public void VsTextViewCreated(IVsTextView textViewAdapter)
        {
            var textView = editorFactory.GetWpfTextView(textViewAdapter);
            if (textView == null) return;

            var generalOptions = Utils.GetGeneralOptionsPage(serviceProvider);
            if (!generalOptions.RenameRefactoringEnabled) return;

            ITextDocument doc;
            if (textDocumentFactoryService.TryGetTextDocument(textView.TextBuffer, out doc))
            {
                if (doc != null)
                    AddCommandFilter(textViewAdapter,
                        new RenameCommandFilter(doc, textView, fsharpVsLanguageService,
                                                serviceProvider, projectFactory));
            }
        }

        private static void AddCommandFilter(IVsTextView viewAdapter, RenameCommandFilter commandFilter)
        {
            if (!commandFilter.IsAdded)
            {
                // Get the view adapter from the editor factory
                IOleCommandTarget next;
                int hr = viewAdapter.AddCommandFilter(commandFilter, out next);

                if (hr == VSConstants.S_OK)
                {
                    commandFilter.IsAdded = true;
                    // You'll need the next target for Exec and QueryStatus
                    if (next != null) commandFilter.NextTarget = next;
                }
            }
        }

    }
}

File: src/FSharpVSPowerTools.Pure/Commands/RenameCommandFilterProvider.fs
module FSharpVSPowerTools.RenameCommandFilterProvider

open System.Diagnostics
open Microsoft.VisualStudio
open Microsoft.VisualStudio.Editor
open Microsoft.VisualStudio.OLE.Interop
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.TextManager.Interop
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open FSharpVSPowerTools.ProjectSystem
open FSharpVSPowerTools.Refactoring
open System.ComponentModel.Composition
open System
open Microsoft.VisualStudio.Text


[<Export (typeof<IVsTextViewCreationListener>)>]
[<ContentType "F#">]
[<TextViewRole (PredefinedTextViewRoles.Editable)>]
type RenameCommandFilterProvider [<ImportingConstructor>]
    ( [<Import (typeof<SVsServiceProvider>)>] 
        serviceProvider             :   IServiceProvider                ,
        textDocumentFactoryService  :   ITextDocumentFactoryService     ,
        editorFacotry               :   IVsEditorAdaptersFactoryService ,
        projectFactory              :   ProjectFactory                  ,
        vsLanguageService           :   VSLanguageService               ) =


    static member AddCommandFilter (viewAdapter:IVsTextView, commandFilter:RenameCommandFilter) =
        if not commandFilter.IsAdded then
            let mutable next = Unchecked.defaultof<IOleCommandTarget>
            let hr = viewAdapter.AddCommandFilter(commandFilter, &next)

            if hr = VSConstants.S_OK then
                commandFilter.IsAdded <- true
                // You'll need the next target for Exec and QueryStatus
                if isNotNull next then commandFilter.NextTarget <- next


    interface IVsTextViewCreationListener with

        member __.VsTextViewCreated textViewAdapter =
            maybe {
                let! textView = editorFacotry.TryGetWpfTextView textViewAdapter
                let! generalOptions = Setting.tryGetGeneralOptions serviceProvider

                if not generalOptions.RenameRefactoringEnabled then return! None else
                let! doc = textDocumentFactoryService.TryDocumentFromBuffer textView.TextBuffer
                RenameCommandFilterProvider.AddCommandFilter ( textViewAdapter, 
                        new RenameCommandFilter( doc, textView, vsLanguageService, serviceProvider, projectFactory))
            } |> ignore

